qiskit/providers/aer/openpulse/duffing_model_generators.py

```diff
@@ -0,0 +1,438 @@
+                               system_list)
+
+
+def _exchange_coupling_terms(coupling_symbols, ordered_edges):
+    """Exchange coupling terms between systems
+
+    Args:
+        coupling_symbols (list): coefficients of exchange couplings
+        ordered_edges (list): list tuples of system indices for the couplings
+    Returns:
+        list: exchange coupling strings
+    """
+
+    idx1_list, idx2_list = zip(*list(ordered_edges))
+
+    return _str_list_generator('2*np.pi*{0}*(Sp{1}*Sm{2}+Sm{1}*Sp{2})',
+                               coupling_symbols,
+                               idx1_list,
+                               idx2_list)
+
+
+def _cr_terms(drive_symbols, driven_system_indices, u_channel_indices):
+    """Cross resonance drive terms
+
+    Args:
+        drive_symbols (list): coefficients for drive terms
+        driven_system_indices (list): list of indices for systems that drive is applied to
+        u_channel_indices (list): indicies for the u_channels corresponding to each term
+    Returns:
+        list: cr term strings
+    """
+
+    return _str_list_generator('2*np.pi*{0}*X{1}||U{2}',
+                               drive_symbols,
+                               driven_system_indices,
+                               u_channel_indices)
+
+
+def _str_list_generator(str_template, *args):
+    """Given a string template, returns a list where each entry is the template formatted by the
+    zip of args. It is assumed that either args is a tuple of lists each of the same length, or
+    is a tuple with each entry beign either an str or int.
+    E.g.
+    1. _str_list_generator('First: {0}, Second: {1}', 'a0', 'b0') returns ['First: a0, Second: b0']
+    2. _str_list_generator('First: {0}, Second: {1}', ['a0', 'a1'], ['b0', 'b1']) returns
+       ['First: a0, Second: b0', 'First: a1, Second: b1']
+
+    Args:
+        str_template (str): string template
+        args (tuple): assumed to be either tuple of iterables of the same length, or a tuple with
+                      entries that are either type str or int
+
+    Returns:
+        list: list of str_template formated by args lists
+    """
+
+    args = [_arg_to_iterable(arg) for arg in args]
+    return [str_template.format(*zipped_arg) for zipped_arg in zip(*args)]
+
+
+def _arg_to_iterable(arg):
+    """Check if arg is an iterable, if not put it into a list. The purpose is to allow arguments
+    of functions to be either lists or singletons, e.g. instead of having to pass ['a'], 'a' can be
+    passed directly.
+
+    Args:
+        arg (Iterable): argument to be checked and turned into an interable if necessary
+
+    Returns:
+        Iterable: either arg, or arg transformed into a list
+    """
+    # catch expected types (issue is str is iterable)
+    if isinstance(arg, (int, str)):
+        return [arg]
+
+    if isinstance(arg, Iterable):
+        return arg
+
+    return [arg]
+
+
+# Helper classes
+
+
+class _coupling_graph:
```

Comment:
Class should be capitalized `class CouplingGraph`

Reply:
I want to keep this private, how is `_CouplingGraph`?